### ISSUE TITLE
Multilabel & non-mutually-exclusive classes

### ISSUE THREAD

**Issue by dithyrambe** [keras-team/keras#8910]
Hi, 
I'm wondering if there is an elegant way to do so :

I have a network outputing a (9 x 9 x 9) matrix (ie. 9 rows by 9 cols by 9 channel, you can see it as a 81 x 9 `Dense` output layer). 
The thing is, I want to output non-mutually-exclusive classes on it. For each 9 by 9 dimension, I need to output the more likely of 9 possible classes (channels). But the 81 "pixels" can share same value.

Here is what I've figured out so far : 

*Nice architecture, but wrong classification task*
```python
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=input_shape))
model.add(Dense(64, activation='relu'))
model.add(Flatten())
model.add(Dense(81*9, activation='softmax'))
model.add(Reshape((9, 9, 9)))

solver.compile(
    optimizer=Adadelta(),
    loss='categorical_crossentropy',  # Here, loss might be the whole problem ...
)
```
The thing is my predictions will sum to 1.0 on each output whereas I want them to sum at 81. This is because optimizer will consider 9 x 9 x 9 mutually-exclusive classes.

*example of output of what I want to avoid*
```
In[]: model.predict(Xtrain).sum((1, 2, 3))  # summing on all dimensions except 0
Out[]: array([1., 1., 1., ..., 1., 1., 1.])  # shape is len(Xtrain) 
```

I manage to workarround with : 

*classificaiton task is fine, but net structure is painful ...*
```python
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=input_shape))
model.add(Dense(64, activation='relu'))
model.add(Flatten())

grid = Input(shape=input_shape)  # define Input Layer to plug on model
features = model(grid)  # define features created by the Network

# define one Dense layer to plug on model
# Each layer is independant
digit_placeholders = [
    Dense(10, activation='softmax')(features)
    for i in range(81)
]

solver = Model(grid, digit_placeholders)

solver.compile(
    optimizer=Adadelta(),
    loss='categorical_crossentropy',  # This time, loss is fine
)
```
This way, each of my 9 x 9 "pixel" (ie 81 output) will look separately for the more likely classes between 9 (channels). This is what I want, however, I don't find it quite elegant ...
Is there a way to achieve the same classification task without defining 81 independents `Dense` output layer. Is there a `loss` function that would help me achieving this ?

Thanks a lot for your help.

dithyrambe

**Comment by gabrieldemarmiesse** [comment]
Hi!
If the channels are your last dimension, you can just call softmax on the last dimension, thus solving your issue. Eg: 
x = Dense(81*9)(x)
x = Reshape((9,9,9))(x)
x = Activation(lamba x: keras.activations.softmax(x,axis=3))(x)

Don't forget that implementation questions should be asked on stackoverflow and don't forget to close the issue if it solves your problem. 

I hope it helps.

**Comment by dithyrambe** [comment]
Oh ! Thanks for the tip ! I should have go on stackoverflow ... my bad.

Thanks a lot, I'll try it tonight ! And I'll close this now. 

Keras is awesome ...

dithyrambe

**Comment by dithyrambe** [comment]
I'm still missing something though. I guess I miss-understood 

Which `loss` should I use then ? I confused myself with activation and loss. 
I think the same trick with the `loss` should get me out of struggle.

Like applying `categorical_crossentropy` on each pixel over the 9 classes. And then optimize the sum of all of them.

I guess there is no `categorical_crossentropy(axis=3)` trick this time :'(

I you have an idea again, I'll be pleased to read you again !

dithyrambe

**Comment by gabrieldemarmiesse** [comment]
Hi!
You were right to worry about it, I read the source code, and while it seems that the keras categorical_crossentropy should work in your case with the backend being tensorflow, it's just pure luck (the sum is performed on the last axis). It won't work in other backends. (And it's not documented to work with 4D tensors in tensorflow, so your code might break in a future keras update).
But you're in luck because in keras it's easy to do custom loss functions! So in your case, you want to compute categorical crossentropy on the last axis, and then average the results over all the other axis. What you can do is:

def categorical_crossentropy_3rd_axis(y_true, y_pred) :
    clipped_y_pred = K.clip(y_pred, K.epsilon(), 1-K.epsilon())
    product = y_true*K.log(clipped_y_pred)
    categorical_crossentropies = K.sum(product, axis=3) #same axis as the softmax
    return K.mean(categorical_crossentropies)

Here you go. This should help.

**Comment by dithyrambe** [comment]
Hello again,

Thanks for you quick, precise and helpful answer !!!

This is so perfect !

However, I guess you somehow forgot the `-` sign somewhere ;).

I end up with : 
```python
def categorical_crossentropy_3rd_axis(y_true, y_pred):
    clipped_y_pred = K.clip(y_pred, K.epsilon(), 1 - K.epsilon())
    product = y_true * K.log(clipped_y_pred)
    categorical_crossentropies = - K.sum(product, axis=3)  # same axis as the softmax
    return K.mean(categorical_crossentropies)
```

Thanks a lot for all your insights !!!

**Comment by gabrieldemarmiesse** [comment]
You are right! I forgot the sign. Good catch.

Glad to help!

**Comment by dithyrambe** [comment]
Hello again,

Actually, I still have some questions if you don't mind ...

Maybe you'll need more context of my use case. I'm working on this Kaggle kernel : https://www.kaggle.com/dithyrambe/neural-nets-as-sudoku-solvers

The purpose of my initial question was to avoid this part.
```python
digit_placeholders = [
    Dense(9, activation='softmax')(features)
    for i in range(81)
]

solver = Model(grid, digit_placeholders)  # build the whole model
```
Thats why I ask my initial question. Thanks to your answer, I manage to get predictions all at once with a single 9 x 9 `softmax` output layer.

Though, the performance is quite lower than the "81 x `Dense` layer" approach. About 99% vs 70% correct grids.

Diving in into `tensorflow`'s implementation of `categorical_crossentropy`, I realize that before clipping, they "normalize output" so they sum to 1 

```python
output /= math_ops.reduce_sum(
        output, axis=len(output.get_shape()) - 1, keep_dims=True)
    # manual computation of crossentropy
    epsilon_ = _to_tensor(epsilon(), output.dtype.base_dtype)
    output = clip_ops.clip_by_value(output, epsilon_, 1. - epsilon_)
    return -math_ops.reduce_sum(
        target * math_ops.log(output),
axis=len(output.get_shape()) - 1)
``` 

So I add to your answer the `output =/ sum(...)` part in the custom objective.

I jump again to 98%.

My questions are these:
- how can "normalisation" to make probabilities sum to 1 can increase performance  ?
- why am I still 1% accuracy under the first approch with 81 x `Dense` layers ? Note this figure is repeatable. Given the same number of epoch, learning rate,  etc ... I'm always a bit less performant with the 9 x 9 `softmax` single layer approch.

Thanks for your insights =)

**Comment by gabrieldemarmiesse** [comment]
Hi!
So multiple things. 

- The tensor that gets out of a softmax activation function is already normalized along the axis of which you've applied the softmax. Since categorical crossentropy expect normalized inputs, keras wants to make sure that you normalized what gets in the categorical crossentropy. In your case, you've used softmax, so you're safe. btw if you do `outputs /= K.sum(outputs, axis=3, keepdims=True)` It'll change nothing the sum will make a tensor of ones. If you didn't specify the axis argument in your sum... Well, you broke categorical crossentropy because now along the 3rd axis, the sum is 1/81.

- There are three main distinct components here. Layers, activations, and loss. With what I gave you, you only applied an activation and a loss. There is no layer, so no chances for your network to improve itself. 
So it's normal that with dense layers it's better. You can always add dense layers to our custom softmax function, it's not incompatible. keras fuses layers and activations because it's handy to write but mathematically, it's totally distinct.

- If you want the dense layers, you can, just don't use any activation when creating the layers if you plan to use softmax just after.

- ProTip: 81 Dense layers are equivalent to one `LocallyConnected2D(9, (1, 1))`. I'll let you read the documentation about it.

I advise you to look at moocs for the difference between layer, activation and loss. fast.ai is pretty good. You have also the stanford class about CV with deep leaning on youtube that is nice.
